### PR TITLE
refactor(compiler): introduce `InterpolationConfig`

### PR DESCRIPTION
**Pull request by lacolaco**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**
- [x] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

`CompilerConfig` takes `interpolation` arguments as `InterpolationConfig` type.

**Other information**:

ref. #8865

`CompilerConfig` has a configuration of interpolation for the parser.
`InterpolationConfig` has two property; `startSymbol` and `endSymbol`.

InterpolationRegexp of the parser will be created with its symbols.

### CODE REVIEW

**Comment by laco0416** [comment]
@vicb I introduced `InterpolationConfig` as the first step to support i18n parser. Please let me know your opinion on this.

cc/ @mhevery

File: modules/@angular/compiler/src/config.ts [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
-              interpolateRegexp: RegExp = null, defaultEncapsulation: ViewEncapsulation = null) {
+              interpolation: InterpolationConfig = null, defaultEncapsulation: ViewEncapsulation = null) {
     if (isBlank(renderTypes)) {
       renderTypes = new DefaultRenderTypes();
     }
     this.renderTypes = renderTypes;
-    if (isBlank(interpolateRegexp)) {
-      interpolateRegexp = DEFAULT_INTERPOLATE_REGEXP;
+    if (isBlank(interpolation)) {
+      interpolation = DAFAULT_INTERPOLATION_CONFIG;
```
Comment (vicb) [inline modules/@angular/compiler/src/config.ts]: DEFAULT

File: modules/@angular/facade/src/lang.ts [hunk 2]
```diff
@@ -506,3 +506,7 @@ export function bitWiseAnd(values: number[]): number {
 export function escape(s: string): string {
   return _global.encodeURI(s);
 }
+
+export function escapeRegExp(s: string) {
```
Comment (vicb) [inline modules/@angular/facade/src/lang.ts]: add return type

File: modules/@angular/facade/src/lang.dart [hunk 3]
```diff
@@ -400,3 +400,7 @@ num bitWiseAnd(List values) {
 String escape(String s) {
   return Uri.encodeComponent(s);
 }
+
+RegExp escapeRegExp(String s) {
```
Comment (vicb) [inline modules/@angular/facade/src/lang.dart]: static method in RegExpWrapper ?

**Comment by vicb** [comment]
@laco0416 LGTM with some minor comments.

Do you plan to amend this PR with i18n / submit a new one ?

**Comment by laco0416** [comment]
@vicb yes. I'll include i18n paraer changes.
2016/06/01 2:00 "Victor Berchet" <notifications@github.com>:

File: modules/@angular/facade/src/lang.dart (hunk 3 above)
Comment (laco0416) [inline modules/@angular/facade/src/lang.dart]: I think that doesn't make sense. RegExpWrapper should be just _Wrapper_ for general RegExp in JS/Dart

**Comment by laco0416** [comment]
- [x] Introduce `InterpolationConfig` to Parser
- [x] Clean up
- [ ] Apply this to HTMLParser / I18nHTMLParser

**Comment by laco0416** [comment]
@vicb It's done. Please review again! :)

**Comment by laco0416** [comment]
Travis red is in saucelabs flaky test.

File: modules/@angular/compiler/src/config.ts [hunk 4]
```diff
@@ -51,6 +52,11 @@ export class DefaultRenderTypes implements RenderTypes {
...
- * A regexp pattern used to interpolate in default.
+ * Interpolation configuration for the parser.
+ * Can be replaced to customize the start/end symbols.
  */
-export var DEFAULT_INTERPOLATE_REGEXP = /\{\{([\s\S]*?)\}\}/g;
+export class InterpolationConfig {
+  constructor(public startSymbol: string, public endSymbol: string) {}
+}
+
+export const DEFAULT_INTERPOLATION_CONFIG = new InterpolationConfig('{{', '}}');
```
Comment (vicb) [inline modules/@angular/compiler/src/config.ts]: Dowe need to `export`, if not you'd better create the obj lazily

File: modules/@angular/compiler/src/i18n/i18n_html_parser.ts [hunk 5]
```diff
@@ -116,9 +117,10 @@ let _PLACEHOLDER_EXPANDED_REGEXP = /<ph(\s)+name=("(\w)+")><\/ph>/gi;
 export class I18nHtmlParser implements HtmlParser {
   errors: ParseError[];
 
-  constructor(private _htmlParser: HtmlParser, private _parser: Parser,
-              private _messagesContent: string, private _messages: {[key: string]: HtmlAst[]},
-              private _implicitTags: string[], private _implicitAttrs: {[k: string]: string[]}) {}
+  constructor(private _config: CompilerConfig, private _htmlParser: HtmlParser,
```
Comment (vicb) [inline modules/@angular/compiler/src/i18n/i18n_html_parser.ts]: Why passing the while config here ?

File: modules/@angular/compiler/src/config.ts (hunk 4 above)
Comment (laco0416) [inline modules/@angular/compiler/src/config.ts]: Totally. :P

File: modules/@angular/facade/src/lang.ts [hunk 6]
```diff
@@ -506,3 +506,7 @@ export function bitWiseAnd(values: number[]): number {
 export function escape(s: string): string {
   return _global.encodeURI(s);
 }
+
+export function escapeRegExp(s: string): string {
+  return StringWrapper.replaceAll(s, /([.*+?^=!:${}()|[\]\/\\])/g, '\\$1');
```
Comment (vicb) [inline modules/@angular/facade/src/lang.ts]: this is TS, no need to go through the wrapper

File: modules/@angular/compiler/src/i18n/i18n_html_parser.ts (hunk 5 above)
Comment (laco0416) [inline modules/@angular/compiler/src/i18n/i18n_html_parser.ts]: I guess I18nHtmlParser will be constructed by Injector in the future, so I set the config token into here. 
But it also can be replaced by `this._parser._config`. Which is better?

File: modules/@angular/compiler/src/config.ts [hunk 7]
```diff
@@ -1,25 +1,26 @@
...
 import {ViewEncapsulation} from '@angular/core';
 
 export class CompilerConfig {
   public renderTypes: RenderTypes;
-  public interpolateRegexp: RegExp;
+  public interpolation: InterpolationConfig;
   public defaultEncapsulation: ViewEncapsulation;
 
   constructor(public genDebugInfo: boolean, public logBindingUpdate: boolean,
-              public useJit: boolean, renderTypes: RenderTypes = null,
```
Comment (vicb) [inline modules/@angular/compiler/src/config.ts]: as we are adding more options, I'm wondering if we should switch to named params. Probably out of scope of this PR though.

Comment (laco0416) [inline modules/@angular/compiler/src/config.ts]: Can I leave that as TODO comment?

Comment (vicb) [inline modules/@angular/compiler/src/config.ts]: Please create an other GH Issue

File: modules/@angular/compiler/src/i18n/i18n_html_parser.ts (hunk 5 above)
Comment (vicb) [inline modules/@angular/compiler/src/i18n/i18n_html_parser.ts]: Makes total sense. Please leave this as is (`_` prefix is supposed to mean `private`)

File: modules/@angular/compiler/src/config.ts (hunk 7 above)
Comment (laco0416) [inline modules/@angular/compiler/src/config.ts]: Done https://github.com/angular/angular/issues/8933

File: modules/@angular/compiler/src/config.ts [hunk 8]
```diff
@@ -51,6 +52,11 @@ export class DefaultRenderTypes implements RenderTypes {
...
- * A regexp pattern used to interpolate in default.
+ * Interpolation configuration for the parser.
+ * Can be replaced to customize the start/end symbols.
  */
-export var DEFAULT_INTERPOLATE_REGEXP = /\{\{([\s\S]*?)\}\}/g;
+export class InterpolationConfig {
+  constructor(public startSymbol: string, public endSymbol: string) {}
+}
+
+const DEFAULT_INTERPOLATION_CONFIG = new InterpolationConfig('{{', '}}');
```
Comment (laco0416) [inline modules/@angular/compiler/src/config.ts]: Can I use `const` here?

File: modules/@angular/compiler/src/expression_parser/parser.ts [hunk 9]
```diff
@@ -65,12 +71,21 @@ export class TemplateBindingParseResult {
...
 
+function _createInterpolateRegExp(config: InterpolationConfig): RegExp {
+  var regexp = escapeRegExp(config.startSymbol) + '([\\s\\S]*?)' + escapeRegExp(config.endSymbol);
+  return RegExpWrapper.create(regexp, 'g');
+}
+
 @Injectable()
 export class Parser {
+  /** @internal */
+  public _interpolateRegExp: RegExp;
```
Comment (vicb) [inline modules/@angular/compiler/src/expression_parser/parser.ts]: do `_interpolateRegExp` &  `_config` need to be public ?

File: modules/@angular/compiler/src/config.ts (hunk 8 above)
Comment (vicb) [inline modules/@angular/compiler/src/config.ts]: yep

**Comment by vicb** [comment]
@laco0416 thanks for your work. I have added a question.

Could you please squash your 2 commits & rebase.

**Comment by vicb** [comment]
thinking more about this, the symbols could not be part of the compiler config but must be part of the `@Component` to be able to mix & match different symbols.

File: modules/@angular/compiler/src/expression_parser/parser.ts (hunk 9 above)
Comment (laco0416) [inline modules/@angular/compiler/src/expression_parser/parser.ts]: no. I'll make them private

**Comment by laco0416** [comment]
@vicb

> thinking more about this, the symbols could not be part of the compiler config but must be part of the `@Component` to be able to mix & match different symbols.

So, the compiler has `{{}}` as a default symbols, and each component has own symbols for _override_ behavior of the compiler... This approach is similar to `encapsulation`, right?

File: modules/@angular/compiler/src/expression_parser/parser.ts (hunk 9 above)
Comment (laco0416) [inline modules/@angular/compiler/src/expression_parser/parser.ts]: Should I make `_lexer` private too?

**Comment by laco0416** [comment]
For that, we must modify the compiler system entirely. I think that is out of this PR... ;(

**Comment by vicb** [comment]
> For that, we must modify the compiler system entirely. I think that is out of this PR... ;(

Either we should revert the partial impl that is currently in master or do the full support (component based).

Would you like to work on that ? We should have the feature soon if we want it in 2.0.0. However it could come in a further release.

**Comment by laco0416** [comment]
Ok, let's revert current impl. you can?
This is not needed as final, I think too.

**Comment by vicb** [comment]
Yes I'll do that.

**Comment by laco0416** [comment]
Thank you. I close here and create new PR as a feature commit.

**Pull request closed without merging**